quantum/audio/audio_chibios_pwm.c

```diff
@@ -0,0 +1,195 @@
+/* Copyright 2020 Jack Humbert
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+Audio Driver: PWM
+
+the duty-cycle is always kept at 50%, and the pwm-period is adjusted to match the frequency of a note to be played back.
+
+this driver uses the chibios-PWM system to produce a squarewave on a given output pin.
+Either in software through a pwm callback and set/clear; or through the pwm hardware which directly toggles the pin via its alternate function. see your MCUs datasheet for pin can be driven by a timers T/IMx_CHy.
+
+ */
+
+
+/* STM32F103C Setup:
+halconf.h:
+#define HAL_USE_PWM                 TRUE
+#define HAL_USE_PAL                 TRUE
+#define HAL_USE_GPT                 TRUE
+
+mcuconf.h:
+#define STM32_PWM_USE_TIM1                  TRUE
+#define STM32_GPT_USE_TIM3                  TRUE
+
+used pin: PA8 (alternate0: Tim1_Ch1)
+
+from the datasheet for STM32F104C8B: alternate function of pin
+TIM1_CH1 = PA8
+TIM1_CH2 = PA9
+TIM1_CH3 = PA10
+TIM1_CH4 = PA11
+
+
+to use another hardware-pwm pin:
+- check the datasheet for the timer_channel and the alternate-function
+- replace PWMDx with the required timer
+- reorder the PCMConfig.channels to match the TIMx_CHy (the stm datasheets usually have CH1 to CH4)
+- set the pin to the appropiate alternate function (beware: differences in chibios GPIOv1, v2, ...)
+*/
```

Comment:
As suggested in one of my other comments, lets parameterise this via `#define`s.
See the configuration items for the I2C driver as an example on how to handle pins, AF modes etc.

Reply:
Under normal circumstances, the Proton-C is what we target when we deal with parameterisation, and the definitions should reflect that. We don't require configurations for every possible combination under the sun, just the main ones. Given that the intended use-case is F103 for the PWM driver, due to the lack of a DAC, I'd say it'd be safe to do something like:
```c
#ifndef DAC_DRIVER
#    ifdef STM32F1xx
         // F103 definitions
#        define DAC_DRIVER DACD1
#    else
         // default F303 definitions
#        define DAC_DRIVER DACD2
#    endif
#endif
```
...or equivalent, for the parameters.

If someone chooses another MCU, the pattern is for whoever the designer is to be able to provide overrides for the `#define`s in their board's `config.h`, allowing for non-invasive changes to be made.

Unfortunately the old ARM audio driver didn't adhere to that pattern, and we've progressively been getting through drivers and turning them into overridable implementations, rather than hard-coded.

In general due to size constraints we tend to prefer the `#define` route rather than weak overridable functions, unless absolutely necessary. It's extremely unlikely we'd need to swap outputs at runtime, so compile-time is preferable here.

Reply:
this commit: 0da097555175b43ae9d20204d4067f289c5ee0eb  exposes the timer, channel and alternate-function to be user configurable - your thoughts on this? :-)